[NhatMinh0208] iP

NhatMinh0208 opened the pull request:
### SpheneBot

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

SpheneBot frees your mind of having to remember things you need to do. It's,


- text-based
- easy to learn
- ~~fast~~ **SUPER FAST** to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. type `todo funtask`
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X]  Managing tasks
- [X]  Managing deadlines
- [ ]  *Reminders (coming soon)*

SpheneBot utilizes the latest techniques in bot programming. Here's an example:
```java

/**
 * Main class of the Sphene bot.
 */
public class Sphene {
    private Storage storage;
    private TaskList tasks;
    private Ui ui;

    /**
     * Creates a new instance of the bot.
     * @param filePath File path to save tasks to.
     */
    public Sphene(String filePath) {
        ui = new Ui();
        storage = new Storage(filePath);
        try {
            tasks = new TaskList(storage.load());
        } catch (SpheneException e) {
            ui.showError(e);
            tasks = new TaskList();
        }
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

NhatMinh0208 committed:
Rename bot to Merylwb

NhatMinh0208 committed:
Implement greet and exit

NhatMinh0208 committed:
Rename bot to Sphene

NhatMinh0208 committed:
Implement echo

NhatMinh0208 committed:
Bring constants out

NhatMinh0208 committed:
Implement ability to add items

NhatMinh0208 committed:
Implement ability to list items

NhatMinh0208 committed:
Add task done status

NhatMinh0208 committed:
Add ability to mark tasks as done / not done

NhatMinh0208 committed:
Add task type ToDo

NhatMinh0208 committed:
Add task type Deadline

NhatMinh0208 committed:
Finish task type Deadline

NhatMinh0208 committed:
Add task type Event

NhatMinh0208 committed:
Add parsing for individual task types

NhatMinh0208 committed:
Fix event icon

NhatMinh0208 committed:
Add initial testing for all commands

NhatMinh0208 committed:
Add exceptions for syntax errors and unknown command errors

NhatMinh0208 committed:
Add excpetions for empty field errors

NhatMinh0208 committed:
Reorganize code for mark/unmark and add out of range exceptions

NhatMinh0208 committed:
Add testing for new excpetions

NhatMinh0208 committed:
Mark exception field as final

NhatMinh0208 committed:
Add ability to delete tasks

NhatMinh0208 committed:
Add ability to delete tasks

NhatMinh0208 committed:
Add ability to save and load tasks

NhatMinh0208 committed:
Merge branch 'branch-Level-7'

NhatMinh0208 committed:
Convert to LocalDateTime type for date-time arguments

NhatMinh0208 committed:
Add proper datetime parsing

NhatMinh0208 committed:
Merge branch 'branch-Level-8'

NhatMinh0208 committed:
Abstract out UI

NhatMinh0208 committed:
Abstract out parser and commands

NhatMinh0208 committed:
Abstract out task list

NhatMinh0208 committed:
Abstract out task list actions

NhatMinh0208 committed:
Abstract out task serialization

NhatMinh0208 committed:
Abstract out file storage

NhatMinh0208 committed:
Implement commands

NhatMinh0208 committed:
Fix some bugs and update test

NhatMinh0208 committed:
Merge branch 'branch-moreOOP'

NhatMinh0208 committed:
Remove unnecessary imports

NhatMinh0208 committed:
Reorganize code into packages

NhatMinh0208 committed:
Merge branch 'branch-Packages'

NhatMinh0208 committed:
Merge branch 'add-gradle-support'

NhatMinh0208 committed:
Add Gradle support for checkstyle

NhatMinh0208 committed:
Fix packaging

NhatMinh0208 committed:
Add test for AddEventCommand

NhatMinh0208 committed:
Add leap day tests for AddEventCommand

NhatMinh0208 committed:
Add test for AddDeadlineCommand

NhatMinh0208 committed:
Add invalid range check for AddEventCommand

NhatMinh0208 committed:
Add JAR build

NhatMinh0208 committed:
Add javadoc for all public facing items

NhatMinh0208 committed:
Refactor code to follow coding standard

NhatMinh0208 committed:
Add find command

NhatMinh0208 committed:
Add parsing for find command

NhatMinh0208 committed:
Merge branch 'branch-A-CodingStandard'

NhatMinh0208 committed:
Merge branch 'branch-A-JavaDoc'

NhatMinh0208 committed:
Fix final style errors

NhatMinh0208 committed:
Merge branch 'branch-Level-9'

NhatMinh0208 committed:
Fix final style errors

soonami69 commented on src/main/java/sphene/component/Parser.java:
```diff
@@ -0,0 +1,182 @@
+    private static final Pattern PATTERN_EVENT = Pattern.compile(" *(.*)/from *(.*)/to *(.*)");
+    private static final Pattern PATTERN_MARK = Pattern.compile(" *(-?[0-9]+) *");
+    private static final Pattern PATTERN_UNMARK = PATTERN_MARK;
+    private static final Pattern PATTERN_DELETE = PATTERN_MARK;
```
I like how you made sure all your constants were in SCREAMING_SNAKE_CASE :)

soonami69 commented on src/main/java/sphene/Sphene.java:
```diff
@@ -0,0 +1,63 @@
+     */
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
```
`isExit` may read unnaturally? Quite a nit, but perhaps something like `shouldExit` may be a more readable/grammar-accurate name?

soonami69 commented on src/main/java/sphene/command/ListCommand.java:
```diff
@@ -0,0 +1,27 @@
+    /**
+     * Creates a new list command.
+     */
+    public ListCommand() {
```
I like how you maintained the use of Egyptian-style braces, even on an empty body! This really shows your **committment** to upholding __Good Coding Standards!__ ❤️

soonami69 commented on src/main/java/sphene/component/Storage.java:
```diff
@@ -0,0 +1,79 @@
+     * @throws LoadException If file cannot be loaded.
+     */
+    public List<String> load() throws LoadException {
+        try {
```
I like how you separated the different `try-catch` blocks with line breaks. It really makes it much more readable.

soonami69 commented on src/main/java/sphene/component/TaskList.java:
```diff
@@ -0,0 +1,177 @@
+    public TaskList find(String query) {
+        TaskList result = new TaskList();
+        tasks.stream()
+                .filter((task) -> task.getContent().contains(query))
```
I like how you added a line break to prevent the line from getting too long, and made sure to properly indent it 8 spaces!

soonami69 commented on src/main/java/sphene/exception/LoadException.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public String getMessage() {
+        return "My dear citizen, I could not retrieve the file " + this.filePath + " from Alexandria's memory.";
```
Maybe consider breaking this line? While not a strict violation of the coding standard (<120 chars), I feel that wrapping could make it a lot more readable.

soonami69 reviewed (commented):
LGTM 👍 Well-formatted code. Very readable and adheres strictly to the coding standards.

NhatMinh0208 committed:
Implement initial GUI and input handler

NhatMinh0208 committed:
Implement GUI for bot output

NhatMinh0208 committed:
Merge branch 'branch-Level-10'

gn07 commented on src/main/java/sphene/component/TaskList.java:
```diff
@@ -0,0 +1,177 @@
+     * @param taskStrings The serialized task strings to be parsed.
+     * @throws TaskLoadFailException If one or more serialized strings cannot be parsed.
+     */
+    public TaskList(List<String> taskStrings) throws TaskLoadFailException {
```
i love how clearly named your methods and exception classes are!

gn07 commented on src/main/java/sphene/task/Event.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString()
```
the line length is kept under 120 chars by wrapping the text, which adheres to coding standards :D

gn07 commented on src/main/java/sphene/component/Ui.java:
```diff
@@ -0,0 +1,109 @@
+     * Alerts the user that a command has been executed successfully.
+     * @param c The executed command.
+     */
+    public void showDone(Command c) {
```
i like how the method names are short and easy to understand

gn07 commented on src/main/java/sphene/component/Parser.java:
```diff
@@ -0,0 +1,182 @@
+        }
+    }
+
+    /**
```
i like how detailed the javadoc comment is on this method

gn07 reviewed (commented):
The code is clean, easy to understand and of high quality. 👍

NhatMinh0208 committed:
Add range assertion for Event constructor

Event constructor did not check that `from` is before `to`.

This means that consumers of Event may accidentally add invalid
events.

=> An assertion that checks that `from` is before `to` is being
added.

NhatMinh0208 committed:
Fix outstanding checkstyle issues

The code was violating some checkstyle rules after adding GUI.

This results in lowered code quality.

=> These issues are now fixed.

NhatMinh0208 committed:
Refactor 'from' and 'to' to 'startTime' and 'endTime'

The datetime fields for the Event task type were not nouns.

This may lead to confusion for readers.

=> The fields have been changed to 'startTime' and 'endTime'.

NhatMinh0208 committed:
Fix capitalization on endTIme

NhatMinh0208 committed:
Refactor 'by' to 'deadlineTime'

NhatMinh0208 committed:
Extract submethods from  in Storage

NhatMinh0208 committed:
Add prefix assertion for storage file path

NhatMinh0208 committed:
Refactor Storage and TaskList

NhatMinh0208 committed:
Merge pull request #1 from NhatMinh0208/branch-A-Assertions

Merge `branch-A-Assertions`

NhatMinh0208 committed:
Merge branch 'master' into branch-A-CodeQuality

NhatMinh0208 committed:
Merge pull request #2 from NhatMinh0208/branch-A-CodeQuality

Merge `branch-A-CodeQuality`

NhatMinh0208 committed:
Add CI via GitHub Actions

NhatMinh0208 committed:
Add dummy sort command

NhatMinh0208 committed:
Add chronological sorting to TaskList + command

NhatMinh0208 committed:
Add UI image and set window title

NhatMinh0208 committed:
Add response after executing command

NhatMinh0208 committed:
Fix Javadocs comment for some methods

NhatMinh0208 committed:
Update user image from default

NhatMinh0208 committed:
Update documentation

NhatMinh0208 committed:
Add new demo image (Ui.png)

NhatMinh0208 committed:
Add more examples to user guide

NhatMinh0208 committed:
Sync README.md to match with user guide

NhatMinh0208 committed:
Fix invalid int exception handling

NhatMinh0208 committed:
Fix exception error messages for invalid int
